File: web/src/lib/filters.ts
import type { Card, CardPriority } from './types.ts';
import { tagKey } from './tags.ts';

export type FilterStatus = 'all' | 'todo' | 'done';
export type FilterPriority = 'all' | CardPriority;
export type FilterHabit = 'all' | 'habit' | 'manual';
export type FilterDeadline = 'all' | 'has_deadline' | 'overdue';

export interface CardFilterState {
  status: FilterStatus;
  priority: FilterPriority;
  tags: string[];
  habit: FilterHabit;
  deadline: FilterDeadline;
  color: string;
}

export const DEFAULT_FILTERS: CardFilterState = {
  status: 'all',
  priority: 'all',
  tags: [],
  habit: 'all',
  deadline: 'all',
  color: 'all',
};

export function countActiveFilters(filters: CardFilterState): number {
  let count = 0;
  if (filters.status !== 'all') count++;
  if (filters.priority !== 'all') count++;
  if (filters.tags.length > 0) count += filters.tags.length;
  if (filters.habit !== 'all') count++;
  if (filters.deadline !== 'all') count++;
  if (filters.color !== 'all') count++;
  return count;
}

export function hasActiveFilters(filters: CardFilterState): boolean {
  return countActiveFilters(filters) > 0;
}

export function matchesFilter(
  card: Card,
  filters: CardFilterState,
  now = new Date(),
): boolean {
  if (filters.status === 'todo' && card.done) return false;
  if (filters.status === 'done' && !card.done) return false;

  if (filters.priority !== 'all' && card.priority !== filters.priority) return false;

  if (filters.tags.length > 0) {
    const cardTagKeys = new Set(card.tags.map(tagKey));
    const matchesAll = filters.tags.every((tag) => cardTagKeys.has(tagKey(tag)));
    if (!matchesAll) return false;
  }

  if (filters.habit === 'habit' && !card.habitId) return false;
  if (filters.habit === 'manual' && card.habitId) return false;

  if (filters.deadline === 'has_deadline' && !card.deadlineAt) return false;
  if (filters.deadline === 'overdue') {
    if (!card.deadlineAt || card.done) return false;
    const deadlineTime = new Date(card.deadlineAt).getTime();
    if (deadlineTime >= now.getTime()) return false;
  }

  if (filters.color !== 'all' && card.color !== filters.color) return false;

  return true;
}

File: mobile/lib/filters.dart
import 'api/models.dart';
import 'tags.dart';

enum FilterStatus { all, todo, done }

enum FilterHabit { all, habit, manual }

enum FilterDeadline { all, overdue, hasDeadline }

class CardFilterState {
  const CardFilterState({
    this.status = FilterStatus.all,
    this.priority = 'all',
    this.tags = const {},
    this.habit = FilterHabit.all,
    this.deadline = FilterDeadline.all,
    this.color = 'all',
  });

  final FilterStatus status;
  final String priority; // 'all' | 'none' | 'low' | 'medium' | 'high' | 'urgent'
  final Set<String> tags;
  final FilterHabit habit;
  final FilterDeadline deadline;
  final String color; // 'all' | color

  static const defaultFilters = CardFilterState();

  int get activeCount {
    var count = 0;
    if (status != FilterStatus.all) count++;
    if (priority != 'all') count++;
    if (tags.isNotEmpty) count += tags.length;
    if (habit != FilterHabit.all) count++;
    if (deadline != FilterDeadline.all) count++;
    if (color != 'all') count++;
    return count;
  }

  bool get hasActiveFilters => activeCount > 0;

  bool matches(PlannerCard card, [DateTime? now]) {
    final currentTime = now ?? DateTime.now();

    // Status
    if (status == FilterStatus.todo && card.done) return false;
    if (status == FilterStatus.done && !card.done) return false;

    // Priority
    if (priority != 'all' && card.priority != priority) return false;

    // Tags
    if (tags.isNotEmpty) {
      final cardTagKeys = card.tags.map(tagKey).toSet();
      final matchesAll = tags.every(
        (t) => cardTagKeys.contains(tagKey(t)),
      );
      if (!matchesAll) return false;
    }

    // Habit
    if (habit == FilterHabit.habit && card.habitId == null) return false;
    if (habit == FilterHabit.manual && card.habitId != null) return false;

    // Deadline
    if (deadline == FilterDeadline.hasDeadline && card.deadlineAt == null) {
      return false;
    }
    if (deadline == FilterDeadline.overdue) {
      if (card.deadlineAt == null || card.done) return false;
      final parsed = DateTime.tryParse(card.deadlineAt!);
      if (parsed == null || !parsed.isBefore(currentTime)) return false;
    }

    // Color
    if (color != 'all' && card.color != color) return false;

    return true;
  }

  CardFilterState copyWith({
    FilterStatus? status,
    String? priority,
    Set<String>? tags,
    FilterHabit? habit,
    FilterDeadline? deadline,
    String? color,
  }) {
    return CardFilterState(
      status: status ?? this.status,
      priority: priority ?? this.priority,
      tags: tags ?? this.tags,
      habit: habit ?? this.habit,
      deadline: deadline ?? this.deadline,
      color: color ?? this.color,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CardFilterState &&
          runtimeType == other.runtimeType &&
          status == other.status &&
          priority == other.priority &&
          tags.length == other.tags.length &&
          tags.containsAll(other.tags) &&
          habit == other.habit &&
          deadline == other.deadline &&
          color == other.color;

  @override
  int get hashCode => Object.hash(
    status,
    priority,
    Object.hashAll(tags),
    habit,
    deadline,
    color,
  );
}
